File: pado-rpc/src/main/python/com/netcrest/pado/rpc/client/biz/query_biz.py
'''
Created on Sep 14, 2017

@author: dpark
'''
import json

from com.netcrest.pado.rpc.rpc_shared import RpcShared
from com.netcrest.pado.rpc.util.rpc_util import create_request


class QueryBiz(RpcShared):
    '''
    QueryBiz provides methods to execute queries on the data node.
    '''
    
    _biz_class_name = 'com.netcrest.pado.rpc.client.biz.QueryBiz';
    
    def __init__(self):
        pass
            
    def execute_pql(self, pql, fetch_size=None):
        '''Execute the specified PQL (Pado Query Language.
        
        Args:
            pql: Pado query statement
            fetch_size: Fetch size per page. This is an optional parameter. The default value is configured by the server.
            
        Returns:
            Query results in JSON RPC reply form.
        '''
        jparams = json.loads('{}')
        jparams['pql'] = pql
        if fetch_size != None:
            jparams['fetchSize'] = fetch_size
        jrequest = create_request(self._biz_class_name, 'executePql', jparams)
        return self.rpc.execute(jrequest, 10)
    
    def next_result_set(self, jresult):
        '''Return the next result set of the query result.
        
        Args:
            jresult: The returned result of execute_pql() or this method.
    
        Returns:
            None if the end of the result is reached.
        '''
        if jresult == None:
            return None
        if 'result' not in jresult:
            return None
        result = jresult['result']
        if 'pql' not in result:
            return None
        else:
            pql = result['pql']
        if 'nextBatchIndexOnServer' not in result:
            nextBatchIndexOnServer = -1
        else:
            nextBatchIndexOnServer = result['nextBatchIndexOnServer']
        if nextBatchIndexOnServer < 0:
            return None
        if 'totalSizeOnServer' not in result:
            totalSizeOnServer = -1
        else:
            totalSizeOnServer = result['totalSizeOnServer']
        if nextBatchIndexOnServer >= totalSizeOnServer:
            return None
        
        startIndex = nextBatchIndexOnServer   
        jparams = json.loads('{}')
        jparams['pql'] = pql
        jparams['startIndex'] = startIndex
        if 'fetchSize' in result:
            fetchSize = result['fetchSize']
            jparams['fetchSize'] = fetchSize
        jrequest2 = create_request(self._biz_class_name, 'executePql', jparams)
        return self.rpc.execute(jrequest2, 10)
        
        
        

File: pado-rpc/src/main/java/com/netcrest/pado/rpc/client/biz/QueryBiz.java
package com.netcrest.pado.rpc.client.biz;

import com.netcrest.pado.data.jsonlite.JsonLite;
import com.netcrest.pado.rpc.client.IRpcContext;
import com.netcrest.pado.rpc.mqtt.client.MqttJsonRpcClient;

@SuppressWarnings({ "rawtypes", "unchecked" })
public class QueryBiz
{	
	private IRpcContext rpcContext;
	
	public QueryBiz(IRpcContext rpcContext)
	{
		this.rpcContext = rpcContext;
	}
	
	public JsonLite executePql(String pql)
	{
		return executePql(pql, -1);
	}
	
	public JsonLite executePql(String pql, int fetchSize)
	{
		JsonLite params = new JsonLite();
		params.put("pql", pql);
		if (fetchSize > 0) {
			params.put("fetchSize", fetchSize);
		}
		return MqttJsonRpcClient.getRpcClient().execute(rpcContext, this, "executePql", params, 0);
	}
	
	public JsonLite nextResultSet(JsonLite result)
	{
		if (result == null) {
			return null;
		}
		result = (JsonLite) result.get("result");
		if (result == null) {
			return null;
		}
		String pql = result.getString("pql", null);
		if (pql == null) {
			return null;
		}
		int nextBatchIndexOnServer = result.getInt("nextBatchIndexOnServer", -1);
		if (nextBatchIndexOnServer < 0) {
			return null;
		}
		int totalSizeOnServer = result.getInt("totalSizeOnServer", -1);
		if (nextBatchIndexOnServer >= totalSizeOnServer) {
			return null;
		}
		int fetchSize = result.getInt("fetchSize", -1);
		int startIndex = nextBatchIndexOnServer;
		
		JsonLite params = new JsonLite();
		params.put("pql", pql);
		params.put("startIndex", startIndex);
		if (fetchSize > 0) {
			params.put("fetchSize", fetchSize);
		}
		System.out.println("QueryBiz.nextResultSet(): params=" + params.toJsonString());
		return MqttJsonRpcClient.getRpcClient().execute(rpcContext, this, "executePql", params, 0);	
	}
	
	public JsonLite prevResultSet(JsonLite result)
	{
		if (result == null) {
			return null;
		}
		String pql = result.getString("pql", null);
		if (pql == null) {
			return null;
		}
		int nextBatchIndexOnServer = result.getInt("nextBatchIndexOnServer", -1);
		int totalSizeOnServer = result.getInt("totalSizeOnServer", -1);
		if (nextBatchIndexOnServer >= totalSizeOnServer) {
			return null;
		}
		int fetchSize = result.getInt("fetchSize", 100);
		int startIndex = nextBatchIndexOnServer - 2*fetchSize;
		if (startIndex < 0) {
			startIndex = 0;
		}
		
		JsonLite params = new JsonLite();
		params.put("pql", pql);
		params.put("startIndex", startIndex);
		params.put("fetchSize", fetchSize);
		return MqttJsonRpcClient.getRpcClient().execute(rpcContext, this, "executePql", params, 0);	
	}
}
